### PR TITLE
gh-119213: Be More Careful About _PyArg_Parser.kwtuple Across Interpreters

### PR DESCRIPTION
**Pull request by ericsnowcurrently**
`_PyArg_Parser` holds static global data generated for modules by Argument Clinic.  The `_PyArg_Parser.kwtuple` field is a tuple object, even though it's stored within a static global.  In some cases the tuple is statically allocated and thus it's okay that it gets shared by multiple interpreters.  However, in other cases the tuple is set lazily, allocated from the heap using the active interprepreter at the point the tuple is needed.

This is a problem once that interpreter is destroyed since `_PyArg_Parser.kwtuple` becomes at dangling pointer, leading to crashes.  It isn't a problem if the tuple is allocated under the main interpreter, since its lifetime is bound to the lifetime of the runtime.  The solution here is to temporarily switch to the main interpreter.  The alternative would be to always statically allocate the tuple.

This change also fixes a bug where only the most recent parser was added to the global linked list.

* Issue: gh-119213

### CODE REVIEW

File: Python/getargs.c [hunk 1]
```diff
@@ -2883,6 +2900,7 @@ _PyArg_Fini(void)
     struct _PyArg_Parser *tmp, *s = _PyRuntime.getargs.static_parsers;
     while (s) {
         tmp = s->next;
+        //assert(tmp != s);
```
Comment (1st1) [inline Python/getargs.c]:
```suggestion
```

File: Python/getargs.c [hunk 2]
```diff
@@ -1969,8 +1986,8 @@ _parser_init(void *arg)
     parser->next = _Py_atomic_load_ptr(&_PyRuntime.getargs.static_parsers);
     do {
         // compare-exchange updates parser->next on failure
-    } while (_Py_atomic_compare_exchange_ptr(&_PyRuntime.getargs.static_parsers,
-                                             &parser->next, parser));
+    } while (!_Py_atomic_compare_exchange_ptr(&_PyRuntime.getargs.static_parsers,
```
Comment (1st1) [inline Python/getargs.c:1989]: Is this change intentional? If so, how did it work before? (I haven't looked at `_Py_atomic_compare_exchange_ptr`)

File: Python/getargs.c [hunk 3]
```diff
@@ -1947,7 +1948,23 @@ _parser_init(void *arg)
...
+        PyThreadState *temp_tstate = NULL;
+        if (!_Py_IsMainInterpreter(PyInterpreterState_Get())) {
+            temp_tstate = PyThreadState_New(_PyInterpreterState_Main());
+            if (temp_tstate == NULL) {
+                return -1;
+            }
+            save_tstate = PyThreadState_Swap(temp_tstate);
+        }
         kwtuple = new_kwtuple(keywords, len, pos);
+        if (temp_tstate != NULL) {
```
Comment (1st1) [inline Python/getargs.c:1963]: Is doing this before checking `if (kwtuple == NULL)` the right way?

File: Python/getargs.c (hunk 2 above)
Comment (ericsnowcurrently) [reply Python/getargs.c:1989]: It is intentional.  This is the fix for the secondary bug I mentioned above.  Doing the compare-exchange again after it already succeeded meant it would try one more time and set `parser->next` to `parser`.  That meant only the most recent addition to the linked list was actually in the list, but otherwise this bug was benign.  I had to fix this in order to reproduce the error on main.  The bug doesn't exist on 3.12.

File: Python/getargs.c (hunk 3 above)
Comment (ericsnowcurrently) [reply Python/getargs.c:1963]: Yeah, we want to clean up the temporary tstate regardless of the outcome of `new_kwtuple()`.

File: Python/getargs.c (hunk 1 above)
Comment (ericsnowcurrently) [reply Python/getargs.c]: done

File: Python/getargs.c (hunk 2 above)
Comment (1st1) [reply Python/getargs.c:1989]: oh wow. TBQH there's aren't easy APIs to use (and I can't even reason about them without studying the docs carefully). Thanks for explaining.

**Review by 1st1 — APPROVED** [review]
